src/storage/db/snapshot.rs

```diff
@@ -264,24 +293,73 @@ pub async fn upload_to_s3(
             .to_string();
         let key = format!("{}/{}", upload_dir, file_name);
 
-        info!(key, "Uploading chunk to s3");
-        let byte_stream = ByteStream::from_path(entry.path()).await?;
-        let upload_result = s3_client
-            .put_object()
-            .key(key.clone())
+        let mut file = tokio::fs::File::open(entry.path()).await?;
+        let mut buffer = Vec::new();
+        file.read_to_end(&mut buffer).await?;
+
+        let create_res = s3_client
+            .create_multipart_upload()
             .bucket(snapshot_config.s3_bucket.clone())
-            .body(byte_stream)
+            .key(key.clone())
             .send()
-            .await;
-        if let Err(err) = &upload_result {
-            info!(
-                "Error uploading to s3: {}, key: {}, bucket: {}",
-                DisplayErrorContext(err),
-                key,
-                snapshot_config.s3_bucket
-            );
+            .await?;
+
+        let upload_id = create_res.upload_id().unwrap();
+        let mut parts = Vec::new();
+
+        // 5 MB is the minimum size
+        for (i, chunk) in buffer.chunks(5 * 1024 * 1024).enumerate() {
+            let part_number = (i + 1) as i32;
+            info!(key, part_number, "Uploading snapshot chunk to s3");
+
+            match s3_client
+                .upload_part()
+                .bucket(snapshot_config.s3_bucket.clone())
+                .key(key.clone())
+                .upload_id(upload_id)
+                .part_number(part_number)
+                .body(ByteStream::from(chunk.to_vec()))
+                .send()
+                .await
+            {
+                Err(err) => {
+                    info!(key, "Aborting multipart snapshot upload");
+                    s3_client
+                        .abort_multipart_upload()
+                        .bucket(snapshot_config.s3_bucket.clone())
+                        .key(key.clone())
+                        .upload_id(upload_id)
+                        .send()
+                        .await?;
+                    return Err(SnapshotError::UploadPartError(err));
+                }
+                Ok(upload_res) => {
+                    parts.push(
+                        aws_sdk_s3::types::CompletedPart::builder()
+                            .part_number(part_number)
+                            .e_tag(upload_res.e_tag().unwrap())
+                            .build(),
+                    );
+                }
+            }
         }
-        upload_result?;
+
+        info!(key, "Finished uploading snapshot to s3");
+
+        // Complete Multipart Upload
+        s3_client
+            .complete_multipart_upload()
```

Comment:
If any of the parts fail, do we need to cancel the multipart upload? Does it have implications for future snapshots?

Reply:
Yeah, we call `abort_multipart_upload` above and that should allow future uploads to go ahead.